includes/Admin/Settings.php

```diff
@@ -7,59 +7,227 @@
+					max="<?php echo esc_attr( $current_job->amount ); ?>"
+					value="<?php echo $current_job->processed ? esc_attr( $current_job->processed ) : ''; ?>"
+					style="width: 560px;"
+				>
+					<?php
+					printf(
+						'%d out of %d',
+						esc_html( $current_job->processed ),
+						esc_html( $current_job->amount ),
+					);
+					?>
+				</progress>
+			</div>
+		<?php elseif ( filter_input( INPUT_POST, 'cancel_job' ) ) : ?>
+			<div class="notice notice-error inline-notice is-dismissible" style="margin-left: 0;">
+				<p>Current job canceled.</p>
+			</div>
+		<?php endif; ?>
+
 		<form method="post">
 			<?php wp_nonce_field( 'generate', 'smoothgenerator_nonce' ); ?>
-			<h2>Generate Products</h2>
+			<h2>Generate products</h2>
 			<p>
-				<input type="number" name="num_products_to_generate" value="<?php echo esc_attr( self::DEFAULT_NUM_PRODUCTS ); ?>" min="1" />
-				Number of products to generate.
+				<label for="generate_products_input" class="screen-reader-text">Number of products to generate</label>
+				<input
+					id="generate_products_input"
+					type="number"
+					name="num_products_to_generate"
+					value="<?php echo esc_attr( self::DEFAULT_NUM_PRODUCTS ); ?>"
+					min="1"
+					<?php disabled( $current_job instanceof AsyncJob ); ?>
+				/>
+				<?php
+				submit_button(
+					'Generate',
+					'primary',
+					'generate_products',
+					false,
+					$generate_button_atts
+				);
+				?>
 			</p>
-			<?php submit_button( 'Generate', 'primary', 'generate_products' ); ?>
 
-			<h2>Generate Orders</h2>
+			<h2>Generate orders</h2>
 			<p>
-				<input type="number" name="num_orders_to_generate" value="<?php echo esc_attr( self::DEFAULT_NUM_ORDERS ); ?>" min="1" />
-				Number of orders to generate.
+				<label for="generate_orders_input" class="screen-reader-text">Number of orders to generate</label>
+				<input
+					id="generate_orders_input"
+					type="number"
+					name="num_orders_to_generate"
+					value="<?php echo esc_attr( self::DEFAULT_NUM_ORDERS ); ?>"
+					min="1"
+					<?php disabled( $current_job instanceof AsyncJob ); ?>
+				/>
+				<?php
+				submit_button(
+					'Generate',
+					'primary',
+					'generate_orders',
+					false,
+					$generate_button_atts
+				);
+				?>
 			</p>
-			<?php submit_button( 'Generate', 'primary', 'generate_orders' ); ?>
 
-			<h2>Cancel all scheduled generations</h2>
-			<p>
-			</p>
-			<?php submit_button( 'Cancel all', 'primary', 'cancel_all_generations' ); ?>
+			<?php
+			submit_button(
+				'Cancel current job',
+				'secondary',
+				'cancel_job',
+				true,
+				$cancel_button_atts
+			);
+			?>
 		</form>
 		<?php
+
+		self::heartbeat_script();
+	}
+
+	/**
+	 * Script to run the progress bar.
+	 *
+	 * @return void
+	 */
+	protected static function heartbeat_script() {
+		?>
+		<script>
+			( function( $ ) {
+				var $progress = $( '#smoothgenerator-progress-bar' );
+				var $controls = $( '[id^="generate_"]' );
+				var $cancel   = $( '#cancel_job' );
+
```

Comment:
Bit of a micro-optimization, perhaps, but since we're using jQuery, what do you think of shaving some costs by getting a wrapped document object here:

```suggestion
				const $document = $( document );

```

Then reusing `$document` in place of `$( document )` in the rest of the function?

Reply:
👍 3f2386e